File: Manuel/개인 활동/Class 3/연결 요소의 개수/BFS/11274_BFS.cpp
#include <iostream>
#include <vector>
#include <queue>
#define endl "\n"

using namespace std;

vector<vector<int>> graph;
vector<bool> visited;

void bfs(int start_node)
{
    queue<int> q;
    q.push(start_node);
    visited[start_node] = true;

    while(!q.empty())
    {
        int u = q.front();
        q.pop();
        for(int v : graph[u])
        {
            if(!visited[v])
            {
                visited[v] = true;
                q.push(v);
            }
        }
    }
}

int main()
{
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);

    int N, M;
    cin >> N >> M;

    graph.resize(N + 1);
    visited.resize(N + 1, false);

    for(int i = 0; i < M; i++)
    {
        int u, v;
        cin >> u >> v;
        graph[u].push_back(v);
        graph[v].push_back(u);
    }

    int connected_count = 0;
    for(int i = 1; i <= N; i++)
    {
        if(!visited[i])
        {
            connected_count++;
            bfs(i);
        }
    }

    cout << connected_count << endl;
    
    return 0;
}

File: Manuel/개인 활동/Class 3/연결 요소의 개수/BFS/11274_BFS.py
import sys
from collections import deque

def bfs(graph, visited, start_node):
    q = deque([start_node])
    visited[start_node] = True
    
    while q:
        u = q.popleft()
        for v in graph[u]:
            if not visited[v]:
                visited[v] = True
                q.append(v)

def main():
    input = sys.stdin.readline

    N, M = map(int, input().split())
    graph = [[] for _ in range(N + 1)]
    visited = [False] * (N + 1)

    for _ in range(M):
        u, v = map(int, input().split())
        graph[u].append(v)
        graph[v].append(u)
    
    connected_count = 0

    for i in range(1, N + 1):
        if not visited[i]:
            connected_count += 1
            bfs(graph, visited, i)
    
    print(connected_count)

main()